### PR TITLE
feat(sfcc): add sfcc cors backend

### PR DESCRIPTION
**Pull request by rs-amp**
This PR adds an SFCC backend that corrects directly to their API, rather than going through a proxy. This is possible when allowing the origin of the extension for CORS from the SFCC configuration, and selecting `allow-same-origin` in the extension permissions list. It behaves similarly to the proxy, except it does all data transformation within the backend.

The only difference in extension side configuration is the omission of the `proxyUrl` param and the backend name `sfcc-cors`.

This doesn't replace the proxied SFCC backend for compatibility reasons. You must manually choose the cors backend.

### CODE REVIEW

File: src/backends/SFCCCors.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    const {hits, total} = responseJson;
+    let items = [];
+    const numPages = Math.ceil(total / PAGE_SIZE);
+    const pageSettings = {numPages, curPage: page.curPage, total};
+
+    if (hits) {
+      items = hits.map(hit => ({
+        id: hit.id,
+        name: (hit.name && hit.name.default) ? hit.name.default : null,
+        image: get(hit, 'image.abs_url', null)
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply src/backends/SFCCCors.js]: For now, I've changed `name` to use the same lodash method.

File: src/backends/SFCCCors.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+  async commonSearch(state, query) {
+    const {
+      page,
+      params: { siteId, sfccUrl: endpoint }
+    } = state;
+
+    const token = await this.getAuth(state);
+
+    const PAGE_SIZE = 20;
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply src/backends/SFCCCors.js]: sure, i made it get this from `state`

File: src/backends/SFCCCors.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    const FIVE_MINS = 300000;
+    const expired = new Date().getTime() + FIVE_MINS;
+    return expired > expires;
+  }  
+
+  async getAuth(state) {
+    const {
+      params: { authSecret, authClientId }
+    } = state;
+    const authUrl = "https://account.demandware.com/dw/oauth2/access_token";
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply src/backends/SFCCCors.js]: now configurable as `authUrl` in params, with the demandware url as the default value.

**Comment by rs-amp** [comment]
I added tests for the new backend in the latest commit, as I didn't notice the project had them initially.

File: src/backends/SFCCCors.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import { get } from 'lodash';
+
+export class SFCCCors {
+  constructor(settings) {
+    this.settings = settings;
+    this.tokens = {};
+  }
+
+  tokenExpired(expires) {
+    const FIVE_MINS = 300000;
```
Comment (Benj0s) [inline src/backends/SFCCCors.js:11]: Same as comment on line 46

**Review by Benj0s — APPROVED** [review]

File: src/backends/SFCCCors.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+      body: 'grant_type=client_credentials',
+    };
+
+    const response = await fetch(authUrl, params);
+    if (!response.ok) {
+      throw new Error('Error fetching token', await response.text(), response.statusText);
+    }
+    const token = await response.json();
+
+    const ONE_SECOND_IN_MILLISECONDS = 1000;
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply src/backends/SFCCCors.js:46]: This one won't change, as it's used to convert seconds into milliseconds.

File: src/backends/SFCCCors.js (hunk 4 above)
Comment (rs-amp) [reply src/backends/SFCCCors.js]: Addressed, now called `EXPIRY_BIAS` and has an explanation comment.

**Review by easen-amp — APPROVED** [review]

**Pull request merged**